sklearn/mixture/gaussianmixture.py

```diff
@@ -0,0 +1,685 @@
+
+    # e-step methods
+    @abstractmethod
+    def _estimate_log_weights(self):
+        pass
+
+    @abstractmethod
+    def _estimate_log_likelihood_full(self, X):
+        pass
+
+    @abstractmethod
+    def _estimate_log_likelihood_tied(self, X):
+        pass
+
+    @abstractmethod
+    def _estimate_log_likelihood_diag(self, X):
+        pass
+
+    @abstractmethod
+    def _estimate_log_likelihood_spherical(self, X):
+        pass
+
+    def _estimate_log_likelihood(self, X):
+        estimate_log_likelihood_functions = {
+            "full": self._estimate_log_likelihood_full,
+            "tied": self._estimate_log_likelihood_tied,
+            "diag": self._estimate_log_likelihood_diag,
+            "spherical": self._estimate_log_likelihood_spherical
+        }
+        weighted_log_likelihood = (self._estimate_log_weights() +
+                                   estimate_log_likelihood_functions
+                                   [self.covariance_type](X))
+        return weighted_log_likelihood
+
+    def _estimate_responsibilities(self, X):
+        weighted_log_likelihood = self._estimate_log_likelihood(X)
+        log_probabilities = logsumexp(weighted_log_likelihood, axis=1)
+        responsibilities = np.exp(weighted_log_likelihood -
+                                  log_probabilities[:, np.newaxis])
+        return responsibilities
+
+    def score_samples(self, X):
+        check_is_fitted(self, 'weights_')
+        check_is_fitted(self, 'means_')
+        check_is_fitted(self, 'covars_')
+        X = _check_X(X, self.n_components, self.means_.shape[1])
+
+        weighted_log_likelihood = self._estimate_log_likelihood(X)
+        log_probabilities = logsumexp(weighted_log_likelihood, axis=1)
+        return log_probabilities
+
+    def _initialize_by_kmeans(self, X):
+        labels = cluster.KMeans(n_clusters=self.n_components,
+                                random_state=self.random_state).fit(X).labels_
+        responsibilities = np.zeros((X.shape[0], self.n_components))
+        responsibilities[range(X.shape[0]), labels] = 1
+        return responsibilities
+
+    def _initialize(self, X, init_weights_, init_means_, init_covars_):
+        if (init_weights_ is None or init_means_ is None or
+           init_covars_ is None):
+            if self.verbose > 1:
+                print_('\n\tInitializing parameters ... ', end='')
+
+            # use self.init_params to initialize
+            if self.init_params == 'kmeans':
+                responsibilities = self._initialize_by_kmeans(X)
+            else:
+                # other initialization methods as long as
+                # they return responsibilities
+                responsibilities = np.random.rand(X.shape[0],
+                                                  self.n_components)
+                responsibilities = (responsibilities /
+                                    responsibilities.sum(axis=1)
+                                    [:, np.newaxis])
```

Comment:
I would also put a log here for the initialization of the components, either via uniform random assignments of the samples or via kmeans.